File: hdf-parsers/typescript/flatten.ts
import type { HdfResults, EvaluatedBaseline, EvaluatedRequirement } from '@mitre/hdf-schema';

// ── Public Types ────────────────────────────────────────────

export interface FlattenResult {
  results: HdfResults;
  metadata: FlattenMetadata;
}

export interface FlattenMetadata {
  originalBaselineCount: number;
  flattenedBaselineCount: number;
  merges: BaselineMerge[];
  warnings: string[];
}

export interface BaselineMerge {
  rootBaseline: string;
  absorbedBaselines: string[];
  controlsBefore: number;
  controlsAfter: number;
  pattern: 'deep' | 'wide' | 'hybrid';
}

// ── Internal Helpers ───────────────────────────────────────

/** BFS from root, returns names in top-down order. Cycle-safe. */
function collectTree(root: string, childrenMap: Map<string, string[]>): string[] {
  const order: string[] = [];
  const seen = new Set<string>();
  const queue = [root];
  while (queue.length > 0) {
    const name = queue.shift()!;
    if (seen.has(name)) continue;
    seen.add(name);
    order.push(name);
    for (const child of (childrenMap.get(name) || [])) {
      queue.push(child);
    }
  }
  return order;
}

/** Classify merge pattern based on tree shape */
function detectPattern(root: string, childrenMap: Map<string, string[]>): 'deep' | 'wide' | 'hybrid' {
  const rootChildren = childrenMap.get(root) || [];
  if (rootChildren.length <= 1) {
    return 'deep';
  }
  for (const child of rootChildren) {
    if ((childrenMap.get(child) || []).length > 0) {
      return 'hybrid';
    }
  }
  return 'wide';
}

/** Merge incoming requirement fields onto existing */
function mergeRequirement(
  existing: EvaluatedRequirement,
  incoming: EvaluatedRequirement
): EvaluatedRequirement {
  const result: EvaluatedRequirement = { ...existing };

  // Impact: incoming always wins (required field, always present)
  result.impact = incoming.impact;

  // Results: keep whichever is non-empty (base has them, overlay doesn't)
  if (incoming.results && incoming.results.length > 0) {
    result.results = incoming.results;
  }

  // Code: incoming wins if non-empty
  if (incoming.code && incoming.code.trim() !== '') {
    result.code = incoming.code;
  }

  // Tags: shallow merge (incoming keys override)
  if (incoming.tags) {
    result.tags = { ...existing.tags, ...incoming.tags };
  }

  // Severity: incoming wins if present, else keep existing
  if (incoming.severity !== undefined) {
    result.severity = incoming.severity;
  }

  // EffectiveStatus: incoming wins only if it has results (otherwise its
  // effectiveStatus is a computed artifact from empty results, not intentional).
  // Overlays typically have empty results — the base has the real test results.
  if (incoming.effectiveStatus !== undefined && incoming.results && incoming.results.length > 0) {
    result.effectiveStatus = incoming.effectiveStatus;
  }

  // Descriptions: merge by label (incoming overrides same label)
  if (incoming.descriptions && incoming.descriptions.length > 0) {
    const descMap = new Map(
      (existing.descriptions || []).map(d => [d.label, d])
    );
    for (const d of incoming.descriptions) {
      descMap.set(d.label, d);
    }
    result.descriptions = [...descMap.values()];
  }

  return result;
}

/**
 * Resolve parentBaseline for a baseline.
 * InSpec parent_profile can use depends-name aliases (e.g., 'k8s' instead of
 * 'k8s-node-stig-baseline'). When the value isn't a direct profile name,
 * find who depends on this baseline — that's the actual parent.
 */
function resolveParentBaseline(
  b: EvaluatedBaseline,
  byName: Map<string, EvaluatedBaseline>,
  allBaselines: EvaluatedBaseline[]
): string | undefined {
  if (!b.parentBaseline) return undefined;
  if (byName.has(b.parentBaseline)) return b.parentBaseline;

  // Alias resolution: find the profile whose depends array includes this baseline
  for (const candidate of allBaselines) {
    if (candidate.depends?.some((d) => d.name === b.name)) {
      return candidate.name;
    }
  }
  return undefined; // orphan
}

// ── Public API ──────────────────────────────────────────────

/**
 * Flatten overlay/wrapper baselines in an HDF Results document.
 *
 * Handles:
 * - Deep nesting (overlay chains with shared control IDs via parentBaseline)
 * - Wide nesting (wrapper profiles aggregating independent bases)
 * - Hybrid (both patterns in one document)
 *
 * @param results - Parsed HDF Results (from parseResults() or equivalent)
 * @returns FlattenResult with flattened data and merge metadata
 */
export function flattenOverlays(results: HdfResults): FlattenResult {
  const { baselines } = results;
  const warnings: string[] = [];
  const merges: BaselineMerge[] = [];

  if (!baselines || baselines.length === 0) {
    return {
      results: { ...results, baselines: [] },
      metadata: {
        originalBaselineCount: 0,
        flattenedBaselineCount: 0,
        merges: [],
        warnings: [],
      },
    };
  }

  // Index baselines by name
  const byName = new Map<string, EvaluatedBaseline>();
  for (const b of baselines) {
    if (byName.has(b.name)) {
      warnings.push(`Duplicate baseline name "${b.name}" — later entry overwrites earlier`);
    }
    byName.set(b.name, b);
  }

  // Resolve parentBaseline aliases and build parent map
  const resolvedParent = new Map<string, string | undefined>();
  for (const b of baselines) {
    const parent = resolveParentBaseline(b, byName, baselines);
    resolvedParent.set(b.name, parent);
    if (b.parentBaseline && !parent) {
      warnings.push(
        `Baseline "${b.name}" references nonexistent parent "${b.parentBaseline}"`
      );
    }
  }

  // Build parent → children adjacency using resolved parents
  const childrenMap = new Map<string, string[]>();
  for (const b of baselines) {
    const parent = resolvedParent.get(b.name);
    if (parent) {
      const list = childrenMap.get(parent) || [];
      list.push(b.name);
      childrenMap.set(parent, list);
    }
  }

  // Find roots: no resolved parent
  const roots: string[] = [];
  const visited = new Set<string>();

  for (const b of baselines) {
    if (!resolvedParent.get(b.name)) {
      roots.push(b.name);
    }
  }

  // Mark reachable from roots (iterative DFS to avoid stack overflow on deep trees)
  function markReachable(start: string): void {
    const stack = [start];
    while (stack.length > 0) {
      const name = stack.pop()!;
      if (visited.has(name)) continue;
      visited.add(name);
      const children = childrenMap.get(name);
      if (children) {
        for (const child of children) {
          stack.push(child);
        }
      }
    }
  }
  for (const r of roots) {
    markReachable(r);
  }

  // Detect cycles: unvisited baselines are in cycles
  for (const b of baselines) {
    if (!visited.has(b.name)) {
      warnings.push(`Circular parentBaseline detected involving "${b.name}"`);
      roots.push(b.name);
      markReachable(b.name);
    }
  }

  // Process each root tree
  const flatBaselines: EvaluatedBaseline[] = [];

  for (const rootName of roots) {
    const root = byName.get(rootName)!;
    const treeNames = collectTree(rootName, childrenMap);

    if (treeNames.length === 1) {
      // Standalone baseline — pass through unchanged (preserve depends)
      flatBaselines.push(root);
      continue;
    }

    // Bottom-up order: reverse top-down BFS
    const bottomUp = [...treeNames].reverse();

    // Merge requirements across the tree
    const merged = new Map<string, EvaluatedRequirement>();
    let controlsBefore = 0;
    const absorbed: string[] = [];

    for (const name of bottomUp) {
      const b = byName.get(name)!;
      controlsBefore += b.requirements.length;
      if (name !== rootName) {
        absorbed.push(name);
      }
      for (const req of b.requirements) {
        if (merged.has(req.id)) {
          merged.set(req.id, mergeRequirement(merged.get(req.id)!, req));
        } else {
          merged.set(req.id, { ...req });
        }
      }
    }

    const mergedReqs = [...merged.values()];
    const pattern = detectPattern(rootName, childrenMap);

    merges.push({
      rootBaseline: rootName,
      absorbedBaselines: absorbed,
      controlsBefore,
      controlsAfter: mergedReqs.length,
      pattern,
    });

    const out: EvaluatedBaseline = {
      ...root,
      requirements: mergedReqs,
    };
    delete out.parentBaseline;
    delete out.depends;
    flatBaselines.push(out);
  }

  return {
    results: { ...results, baselines: flatBaselines },
    metadata: {
      originalBaselineCount: baselines.length,
      flattenedBaselineCount: flatBaselines.length,
      merges,
      warnings,
    },
  };
}

File: hdf-parsers/go/flatten.go
package hdfparsers

import (
	"fmt"
	"strings"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go"
)

// ── Public Types ────────────────────────────────────────────

// MergePattern classifies how baselines were related.
type MergePattern string

const (
	PatternDeep   MergePattern = "deep"
	PatternWide   MergePattern = "wide"
	PatternHybrid MergePattern = "hybrid"
)

// FlattenResult is the output of FlattenOverlays.
type FlattenResult struct {
	Results  hdf.HDFResults  `json:"results"`
	Metadata FlattenMetadata `json:"metadata"`
}

// FlattenMetadata describes what was merged.
type FlattenMetadata struct {
	OriginalBaselineCount  int             `json:"originalBaselineCount"`
	FlattenedBaselineCount int             `json:"flattenedBaselineCount"`
	Merges                 []BaselineMerge `json:"merges"`
	Warnings               []string        `json:"warnings"`
}

// BaselineMerge describes a single merge operation.
type BaselineMerge struct {
	RootBaseline      string       `json:"rootBaseline"`
	AbsorbedBaselines []string     `json:"absorbedBaselines"`
	ControlsBefore    int          `json:"controlsBefore"`
	ControlsAfter     int          `json:"controlsAfter"`
	Pattern           MergePattern `json:"pattern"`
}

// ── Internal Helpers ───────────────────────────────────────

// collectTree does cycle-safe BFS from root, returning names in top-down order.
func collectTree(root string, childrenMap map[string][]string) []string {
	var order []string
	seen := map[string]bool{}
	queue := []string{root}
	for len(queue) > 0 {
		name := queue[0]
		queue = queue[1:]
		if seen[name] {
			continue
		}
		seen[name] = true
		order = append(order, name)
		queue = append(queue, childrenMap[name]...)
	}
	return order
}

// detectPattern classifies the tree shape.
func detectPattern(root string, childrenMap map[string][]string) MergePattern {
	rootChildren := childrenMap[root]
	if len(rootChildren) <= 1 {
		return PatternDeep
	}
	for _, child := range rootChildren {
		if len(childrenMap[child]) > 0 {
			return PatternHybrid
		}
	}
	return PatternWide
}

// mergeRequirement merges incoming requirement fields onto existing.
func mergeRequirement(existing, incoming hdf.EvaluatedRequirement) hdf.EvaluatedRequirement {
	result := existing // copy

	// Impact: incoming always wins
	result.Impact = incoming.Impact

	// Results: keep whichever is non-empty
	if len(incoming.Results) > 0 {
		result.Results = incoming.Results
	}

	// Code: incoming wins if non-empty
	if incoming.Code != nil && strings.TrimSpace(*incoming.Code) != "" {
		result.Code = incoming.Code
	}

	// Tags: shallow merge (incoming keys override)
	if incoming.Tags != nil {
		merged := make(map[string]interface{})
		for k, v := range existing.Tags {
			merged[k] = v
		}
		for k, v := range incoming.Tags {
			merged[k] = v
		}
		result.Tags = merged
	}

	// Severity: incoming wins if present, else keep existing
	if incoming.Severity != nil {
		result.Severity = incoming.Severity
	}

	// EffectiveStatus: incoming wins only if it has results (otherwise its
	// EffectiveStatus is a computed artifact from empty results, not intentional).
	// Overlays typically have empty results — the base has the real test results.
	if incoming.EffectiveStatus != nil && len(incoming.Results) > 0 {
		result.EffectiveStatus = incoming.EffectiveStatus
	}

	// Descriptions: merge by label with deterministic order (existing first, then new from incoming)
	if len(incoming.Descriptions) > 0 {
		descMap := make(map[string]hdf.Description, len(existing.Descriptions)+len(incoming.Descriptions))
		orderedLabels := make([]string, 0, len(existing.Descriptions)+len(incoming.Descriptions))

		for _, d := range existing.Descriptions {
			if _, seen := descMap[d.Label]; !seen {
				orderedLabels = append(orderedLabels, d.Label)
			}
			descMap[d.Label] = d
		}
		for _, d := range incoming.Descriptions {
			if _, seen := descMap[d.Label]; !seen {
				orderedLabels = append(orderedLabels, d.Label)
			}
			descMap[d.Label] = d
		}

		descs := make([]hdf.Description, 0, len(orderedLabels))
		for _, label := range orderedLabels {
			descs = append(descs, descMap[label])
		}
		result.Descriptions = descs
	}

	return result
}

// resolveParentBaseline resolves a baseline's parentBaseline value.
// InSpec parent_profile can use depends-name aliases. When the value isn't
// a direct profile name, find who depends on this baseline — that's the parent.
func resolveParentBaseline(b hdf.EvaluatedBaseline, byName map[string]hdf.EvaluatedBaseline, all []hdf.EvaluatedBaseline) *string {
	if b.ParentBaseline == nil {
		return nil
	}
	parent := *b.ParentBaseline
	if _, ok := byName[parent]; ok {
		return &parent
	}
	// Alias resolution: find who depends on this baseline
	for _, candidate := range all {
		for _, dep := range candidate.Depends {
			if dep.Name != nil && *dep.Name == b.Name {
				name := candidate.Name
				return &name
			}
		}
	}
	return nil // orphan
}

// ── Public API ──────────────────────────────────────────────

// FlattenOverlays flattens overlay/wrapper baselines in an HDF Results document.
//
// Handles:
//   - Deep nesting (overlay chains with shared control IDs via ParentBaseline)
//   - Wide nesting (wrapper profiles aggregating independent bases)
//   - Hybrid (both patterns in one document)
func FlattenOverlays(results hdf.HDFResults) FlattenResult {
	baselines := results.Baselines
	var warnings []string
	var merges []BaselineMerge

	if len(baselines) == 0 {
		return FlattenResult{
			Results: results,
			Metadata: FlattenMetadata{
				Merges:   []BaselineMerge{},
				Warnings: []string{},
			},
		}
	}

	// Index by name
	byName := make(map[string]hdf.EvaluatedBaseline, len(baselines))
	for _, b := range baselines {
		if _, exists := byName[b.Name]; exists {
			warnings = append(warnings,
				fmt.Sprintf("Duplicate baseline name %q — later entry overwrites earlier", b.Name))
		}
		byName[b.Name] = b
	}

	// Resolve parentBaseline aliases
	resolvedParent := make(map[string]*string, len(baselines))
	for _, b := range baselines {
		resolved := resolveParentBaseline(b, byName, baselines)
		resolvedParent[b.Name] = resolved
		if b.ParentBaseline != nil && resolved == nil {
			warnings = append(warnings,
				fmt.Sprintf("Baseline %q references nonexistent parent %q", b.Name, *b.ParentBaseline))
		}
	}

	// Build parent → children adjacency
	childrenMap := make(map[string][]string)
	for _, b := range baselines {
		parent := resolvedParent[b.Name]
		if parent != nil {
			childrenMap[*parent] = append(childrenMap[*parent], b.Name)
		}
	}

	// Find roots
	var roots []string
	visited := map[string]bool{}

	for _, b := range baselines {
		if resolvedParent[b.Name] == nil {
			roots = append(roots, b.Name)
		}
	}

	// Mark reachable from roots (iterative DFS to avoid stack overflow on deep trees)
	markReachable := func(start string) {
		stack := []string{start}
		for len(stack) > 0 {
			name := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if visited[name] {
				continue
			}
			visited[name] = true
			stack = append(stack, childrenMap[name]...)
		}
	}
	for _, r := range roots {
		markReachable(r)
	}

	// Detect cycles: unvisited baselines
	for _, b := range baselines {
		if !visited[b.Name] {
			warnings = append(warnings,
				fmt.Sprintf("Circular parentBaseline detected involving %q", b.Name))
			roots = append(roots, b.Name)
			markReachable(b.Name)
		}
	}

	// Process each root tree
	var flatBaselines []hdf.EvaluatedBaseline

	for _, rootName := range roots {
		root := byName[rootName]
		treeNames := collectTree(rootName, childrenMap)

		if len(treeNames) == 1 {
			// Standalone — pass through unchanged (preserve depends)
			flatBaselines = append(flatBaselines, root)
			continue
		}

		// Bottom-up: reverse top-down BFS
		bottomUp := make([]string, len(treeNames))
		for i, name := range treeNames {
			bottomUp[len(treeNames)-1-i] = name
		}

		// Merge requirements
		merged := make(map[string]hdf.EvaluatedRequirement)
		var mergeOrder []string // track insertion order
		controlsBefore := 0
		var absorbed []string

		for _, name := range bottomUp {
			b := byName[name]
			controlsBefore += len(b.Requirements)
			if name != rootName {
				absorbed = append(absorbed, name)
			}
			for _, req := range b.Requirements {
				if existing, ok := merged[req.ID]; ok {
					merged[req.ID] = mergeRequirement(existing, req)
				} else {
					merged[req.ID] = req
					mergeOrder = append(mergeOrder, req.ID)
				}
			}
		}

		mergedReqs := make([]hdf.EvaluatedRequirement, 0, len(merged))
		for _, id := range mergeOrder {
			mergedReqs = append(mergedReqs, merged[id])
		}

		pattern := detectPattern(rootName, childrenMap)

		merges = append(merges, BaselineMerge{
			RootBaseline:      rootName,
			AbsorbedBaselines: absorbed,
			ControlsBefore:    controlsBefore,
			ControlsAfter:     len(mergedReqs),
			Pattern:           pattern,
		})

		out := root
		out.Requirements = mergedReqs
		out.ParentBaseline = nil
		out.Depends = nil
		flatBaselines = append(flatBaselines, out)
	}

	// Preserve all non-baseline fields
	flatResults := results
	flatResults.Baselines = flatBaselines

	if warnings == nil {
		warnings = []string{}
	}
	if merges == nil {
		merges = []BaselineMerge{}
	}

	return FlattenResult{
		Results: flatResults,
		Metadata: FlattenMetadata{
			OriginalBaselineCount:  len(baselines),
			FlattenedBaselineCount: len(flatBaselines),
			Merges:                 merges,
			Warnings:               warnings,
		},
	}
}
